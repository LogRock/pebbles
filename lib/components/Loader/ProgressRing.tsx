import React, { FC, useContext } from "react";

import styled, { ThemeContext } from "styled-components";
import { TypographyParagraphProps } from "../Typography";

import { ProgressRingProps } from "./Loader.types";

export const RingWrapper = styled.div<{ size: number }>`
  position: relative;
  width: ${(props: any) => props.size}px;
  height: ${(props: any) => props.size}px;

  svg {
    position: absolute;
    top: 50%;
    left: 50%;

    transform: translate(-50%, -50%);
  }

  circle {
    transform: rotate(-90deg);
    transform-origin: 50% 50%;
    transition: stroke-dashoffset 0.35s;
  }
`;

export const Label = styled.div<TypographyParagraphProps>`
  position: absolute;
  top: 50%;
  left: 50%;

  transform: translate(-50%, -50%);

  color: ${({ theme }) => theme.colors.neutral["900"]};
  font-family: ${({ theme }) => theme.typography.paragraphXSmall.fontFamily};
  font-size: ${({ theme }) => theme.typography.paragraphXSmall.fontSize};
  font-style: normal;
  font-weight: ${({ theme, weight }) =>
    weight
      ? theme.typography.paragraphXSmall.weights[weight]
      : theme.typography.paragraphXSmall.weights.normal};
  line-height: ${({ theme }) => theme.typography.paragraphXSmall.lineHeight};
`;

const ProgressRing: FC<ProgressRingProps> = ({
  children,
  radius,
  stroke,
  percentage,
  colorGradient,
  showBaseRing,
  label,
  ...inputProps
}) => {
  const theme = useContext(ThemeContext);

  const normalizedRadius = radius - stroke * 2;
  const circumference = normalizedRadius * 2 * Math.PI;
  const strokeDashoffset = circumference - (percentage / 100) * circumference;

  const getRingColor = () => {
    return colorGradient ? "url(#ring_gradient)" : theme.colors.primary[500];
  };

  return (
    <RingWrapper size={radius * 2}>
      {showBaseRing && (
        <svg height={radius * 2} width={radius * 2}>
          <circle
            stroke={theme.colors.primary[50]}
            fill="transparent"
            strokeWidth={stroke}
            style={{ strokeDashoffset }}
            r={normalizedRadius}
            cx={radius}
            cy={radius}
          />
        </svg>
      )}
      <svg height={radius * 2} width={radius * 2}>
        <defs>
          <linearGradient
            id="ring_gradient"
            x1="0"
            y1="0"
            x2="71.8942"
            y2="10.6352"
            gradientUnits="userSpaceOnUse"
          >
            <stop stopColor="#FF5F6D" />
            <stop offset="1" stopColor="#FFC371" />
          </linearGradient>
        </defs>
        <circle
          stroke={getRingColor()}
          fill="transparent"
          strokeWidth={stroke}
          strokeDasharray={circumference + " " + circumference}
          style={{ strokeDashoffset }}
          r={normalizedRadius}
          cx={radius}
          cy={radius}
        />
      </svg>
      {label && <Label>{label}</Label>}
    </RingWrapper>
  );
};

export default ProgressRing;
