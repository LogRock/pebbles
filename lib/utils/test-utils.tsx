/* eslint-disable import/export */
import React from "react";
import { cleanup, render } from "@testing-library/react";
import { afterEach } from "vitest";
import { ThemeProvider } from "styled-components";
import { light as theme } from "../theme";

afterEach(() => {
  cleanup();
});

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => (
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    ),
    ...options,
  });

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
// override render export
export { customRender as render };
