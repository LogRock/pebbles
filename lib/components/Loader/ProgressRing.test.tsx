import React from "react";
import { render, screen } from "../../utils/test-utils";
import ProgressRing from "./ProgressRing";

describe("ProgressRing tests", () => {
  it("matches the component snapshot", () => {
    const { container } = render(
      <ProgressRing
        percentage={75}
        radius={30}
        stroke={4}
        label="75%"
      ></ProgressRing>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows the number percentage", () => {
    render(
      <ProgressRing
        percentage={75}
        radius={30}
        stroke={4}
        label="75%"
      ></ProgressRing>
    );

    expect(screen.getByText("75%")).toBeInTheDocument();
  });
});
