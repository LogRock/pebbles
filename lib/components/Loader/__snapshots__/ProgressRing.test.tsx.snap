// Vitest Snapshot v1

exports[`ProgressRing tests > matches the component snapshot 1`] = `
<div
  class="sc-bcXHqe bMQecf"
  size="60"
>
  <svg
    height="60"
    width="60"
  >
    <defs>
      <lineargradient
        gradientUnits="userSpaceOnUse"
        id="ring_gradient"
        x1="0"
        x2="71.8942"
        y1="0"
        y2="10.6352"
      >
        <stop
          stop-color="#FF5F6D"
        />
        <stop
          offset="1"
          stop-color="#FFC371"
        />
      </lineargradient>
    </defs>
    <circle
      cx="30"
      cy="30"
      fill="transparent"
      r="22"
      stroke="#F1B72C"
      stroke-dasharray="138.23007675795088 138.23007675795088"
      stroke-width="4"
      style="stroke-dashoffset: 34.55751918948772;"
    />
  </svg>
  <div
    class="sc-gswNZR gydReF"
  >
    75%
  </div>
</div>
`;
